### DISCUSSION TITLE
Difference between initializing Firebase app at Native vs JS side?

### DISCUSSION THREAD

**Discussion by TommyLeong** [invertase/react-native-firebase#4465] [Q&A]
Hello all. May I know if there're any differences between initializing Firebase app at Native vs JS side?

One thing I realized is the option config that is passed to  (JS) `initializeApp()` and (Native iOS)`configure()` has slight different in terms of number of information.

Native side (iOS)
```
  if ([FIRApp defaultApp] == nil) {
    [FIRApp configure];
  }
```

JS Side
```
const iosConfig = {
    clientId: 'xx',
    appId: 'xx',
    apiKey: 'xx',
    databaseURL: 'xx',
    storageBucket: 'xx',
    messagingSenderId: 'xx',
    projectId: 'xx',
  };
firebase.initializeApp(iosConfig);
```

**Comment by mikehardy** [comment]
🤔 

> One thing I realized is the option config that is passed to (JS) initializeApp() and (Native iOS)configure() has slight different in terms of number of information.

Please be specific, do you instrument / output the options immediately before they are passed into the underlying firebase app construction? If not do so and highlight the differences so they may be investigated, then for each difference examine the default value to see if any "extra" item is just the default, meaning it is extra but unimportant

**Reply by TommyLeong** [reply]
I do print the "options" out before they are passed into the `configure` method. Well, in Xcode the options are getting all the keys:values from `GoogleService-Info` that is within the project. 
![image](https://user-images.githubusercontent.com/9990179/97459211-7e41aa00-1976-11eb-9d79-80ea369f8437.png)


However as I have pointed out, at JS side.. I realized we initialize the Firebase app with the code below
```
const iosConfig = {
    clientId: 'xx',
    appId: 'xx',
    apiKey: 'xx',
    databaseURL: 'xx',
    storageBucket: 'xx',
    messagingSenderId: 'xx',
    projectId: 'xx',
  };
firebase.initializeApp(iosConfig);
```

**Reply by mikehardy** [reply]
What exactly is the problem? Is there a problem?

**Reply by TommyLeong** [reply]
Remember the discussion we had earlier about "new FCM Token" upon app launch (after app killed)? I realized if we initialize the Firebase App at JS side instead at Native level, the token stabilize. Vice versa, if I do it at Native level.. the issue occur again.

**Reply by mikehardy** [reply]
I do remember that but thought you tracked it to a 3rd party library? Either way, that definitely motivates figuring out what. Then I return to the question of exactly what is in the data structure passed to the initializer from each style, and for each difference what is the default value compared with what is sent when it is there (versus missing). That will allow an analysis of 1) whether there is a semantic configuration difference (the case where things are included sometimes, with non default values) and then each can be tracked

**Reply by TommyLeong** [reply]
You mean to say, whatever is missing values are the 2nd method (JS) as compared to 1st (Native), the default values for these keys might be the reason?

**Reply by mikehardy** [reply]
that is how I would investigate it. If there is a difference, then there is *likely* a difference in contents of data structure used to init. If there is a difference in contents on data structure used to init then each difference should be examined.

If this hypothesis is incorrect and there is no difference or the differences still reduce to both just being default value (whether explicitly configured or implicitly the default) then we will have to re-think. It could be an order of operations problem at that point

**Reply by TommyLeong** [reply]
Thanks for the input. If I were to continue investigate on this, my question is.. how do I examined the differences among all these different input? (I'm cluesless on what should be checked)

Perhaps _a little irrelevant from my topic_, to your knowledge .. do you think it's possible to have 2 SDKs (RNFirebase SDK & 3rd Party SDK) perform Firebase app init at the same time? Which caused Firebase Server to return different token on each app launch?

**Reply by mikehardy** [reply]
That sounds like a terrible idea, two things attempting to control firebase at the same time. I wouldn't contemplate it. For diagnosis I would reach right into node_modules, then because my objective-c skills are terrible I would search on stack-overflow or internet at large how to print data structures to console (I've done it before but have to re-check every time...) and then I would add code before init that prints out the options. Then I would compare them as I try the native or JS init

**Reply by TommyLeong** [reply]
I didn't get you, when you say "data structures", what exactly are you trying to find from the code?

**Reply by mikehardy** [reply]
I believe there is a FirebaseOptions object passed in to the iniitalizer. That is a data structure - an object holding data - and you can inspect it and log out all of it's contents prior to calling the API
